--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -10,6 +10,7 @@
 #include <linux/file.h>
 #include <linux/uio.h>
 #include <linux/fsnotify.h>
+#include <linux/kernelsu.h>
 #include <linux/security.h>
 #include <linux/export.h>
 #include <linux/syscalls.h>
@@ -432,6 +433,11 @@
 {
 	ssize_t ret;
 
+	/* KernelSU-Next Hook */
+	extern int ksu_handle_vfs_read(struct file **file_ptr, char __user **buf_ptr, size_t *count_ptr, loff_t **pos);
+	ksu_handle_vfs_read(&file, &buf, &count, &pos);
+
 	if (!(file->f_mode & FMODE_READ))
 		return -EBADF;
 	if (!(file->f_mode & FMODE_CAN_READ))
@@ -493,6 +499,11 @@
 {
 	ssize_t ret;
 
+	/* KernelSU-Next Hook */
+	extern int ksu_handle_vfs_write(struct file **file_ptr, const char __user **buf_ptr, size_t *count_ptr, loff_t **pos);
+	ksu_handle_vfs_write(&file, &buf, &count, &pos);
+
 	if (!(file->f_mode & FMODE_WRITE))
 		return -EBADF;
 	if (!(file->f_mode & FMODE_CAN_WRITE))

--- a/fs/exec.c
+++ b/fs/exec.c
@@ -55,6 +55,7 @@
 #include <linux/compiler.h>
 #include <linux/oom.h>
 #include <linux/compat.h>
+#include <linux/kernelsu.h>
 
 #include <asm/uaccess.h>
 #include <asm/mmu_context.h>
@@ -1659,6 +1660,10 @@
 	struct linux_binprm *bprm;
 	int retval;
 
+	/* KernelSU-Next Hook */
+	extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv, void *envp, int *flags);
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+
 	BUG_ON(flags & ~AT_SYMLINK_NOFOLLOW);
 
 	retval = unshare_files(&displaced);

--- a/fs/open.c
+++ b/fs/open.c
@@ -32,6 +32,7 @@
 #include <linux/falloc.h>
 #include <linux/fs_struct.h>
 #include <linux/ima.h>
+#include <linux/kernelsu.h>
 #include <linux/dpt_i2o.h>
 
 #include "internal.h"
@@ -1071,6 +1072,10 @@
 {
 	struct file *f;
 
+	/* KernelSU-Next Hook */
+	extern int ksu_handle_faccessat(int *dfd, const char __user **filename_ptr, int *mode, int *flags);
+	ksu_handle_faccessat(&dfd, &filename, &mode, &flags);
+
 	return do_faccessat(dfd, filename, mode);
 }
